### ISSUE TITLE
custom data tests with CTEs fail

### ISSUE THREAD

**Issue by swanderz** [dbt-msft/dbt-synapse#25]
## background
Per @alittlesliceoftom's [Slack thread](https://getdbt.slack.com/archives/C2JRRQDTL/p1604575110440800).

## Steps to reproduce:
1. write a data test that uses CTE clause,
2. dbt wraps the data test's query into another CTE.
3. Synapse returns an error because it does not allow for nested CTEs.

### `tests/example.sql`
```sql
WITH cte_test AS (
    SELECT * FROM {{ref('interesting_moments')}}
)
select TOP 0 * FROM cte_test
```
### `target/compiled/ava_its_data/tests/example.sql`
This doesn't work as Synapse does not allow a single `WITH` clause inside of 
```sql
with dbt__CTE__INTERNAL_test as (
WITH cte_test AS (
    SELECT * FROM "june_sql_pool"."ajs_mkt"."interesting_moments"
)
select TOP 0 * FROM cte_test
)select count(*) from dbt__CTE__INTERNAL_test
```

### create and dispose of a view instead of a CTE
```sql
CREATE VIEW  dbt__CTE__INTERNAL_test as
WITH cte_test AS (
    SELECT * FROM "june_sql_pool"."ajs_mkt"."interesting_moments"
)
select TOP 0 * FROM cte_test;

select count(*) from dbt__CTE__INTERNAL_test

DROP VIEW dbt__CTE__INTERNAL_test
```
### create and dispose of a temp table instead of a CTE?
haven't thought much about the best solution yet...

**Comment by swanderz** [comment]
@jtcohen6 I looked at https://github.com/fishtown-analytics/dbt/pull/2712, but I'm still a little lost as to how (or where) to do the overriding. I got very excited when I saw `dbt-spark` had the same issue (https://github.com/fishtown-analytics/dbt-spark/issues/11) but it looks like a Spark update obviated the need for overwriting. Perhaps if we wait a year, [the same thing](https://feedback.azure.com/forums/307516-azure-synapse-analytics/suggestions/14876727-support-for-recursive-cte) will happen for Synapse?

**Comment by swanderz** [comment]
this is the same issue as https://github.com/mikaelene/dbt-sqlserver/issues/26

**Comment by jtcohen6** [comment]
@swanderz I had a chance to mess around with this, check out #29. I tried implementing your second idea for potential workarounds.

**Comment by alittlesliceoftom** [comment]
Hi both, 

This issue maybe should be broadened to "Logic dependent on nested CTE's doesn't work". 

Another example in the estate is in dbt-utils: 

{{ dbt_utils.date_spine(
    datepart="day",
    start_date="cast(GETDATE() -10 AS DATE)",
    end_date="cast(GETDATE() as date)"
   )
}} 

creates: 


/*
call as follows:

date_spine(
    "day",
    "to_date('01/01/2016', 'mm/dd/yyyy')",
    "dateadd(week, 1, current_date)"
)

*/

with rawdata as (

    

    

    with p as (
        select 0 as generated_number union all select 1
    ), unioned as (

    select

    
    p0.generated_number * pow(2, 0)
     + 
    
    p1.generated_number * pow(2, 1)
     + 
    
    p2.generated_number * pow(2, 2)
     + 
    
    p3.generated_number * pow(2, 3)
    
    
    + 1
    as generated_number

    from

    
    p as p0
     cross join 
    
    p as p1
     cross join 
    
    p as p2
     cross join 
    
    p as p3
    
    

    )

    select *
    from unioned
    where generated_number <= 10
    order by generated_number



),

all_periods as (

    select (
        
  

    dateadd(
        day,
        row_number() over (order by 1) - 1,
        cast(GETDATE() -10 AS DATE)
        )



    ) as date_day
    from rawdata

),

filtered as (

    select *
    from all_periods
    where date_day <= cast(GETDATE() as date)

)

select * from filtered

Of course this one we could hopefully fix in our tsql-utils package.

**Comment by swanderz** [comment]
good point, but I think you're right in that we'll have to solve this problem in two places, in the `__add_ctes()` method, and in `tsql-utils`
